feat: add `name`, introduces struct `ConstructorParams`

smol-ninja opened the pull request:
closes #233 

Adding `name` throws `stack too deep` error because of the large number of variables in function `createMerkleStreamerLL`. In order to fix that, this PR introduces a new struct `ConstructorParams` for `SablierV2MerkleStreamer` constructor parameters.

```solidity
struct ConstructorParams {
    address initialAdmin;
    IERC20 asset;
    string name;
    bytes32 merkleRoot;
    uint40 expiration;
    bool cancelable;
    bool transferable;
}
```

## Summary

### Source
- uses struct `ConstructorParams` to encapsulate `SablierV2MerkleStreamer` constructor parameters.
- adds `name` parameter of type string.
- adds a new internal storage `NAME` of type `bytes32` to store campaign name. `bytes32` is preferred over `string` to optimize for gas.
- adds a public function `name()` to retrieves the campaign name as `string`.

### Events
- update event `CreateMerkleStreamerLL`.
- restructures the order of parameters.

### Errors
- adds a new error `SablierV2MerkleStreamer_CampaignNameTooLong`.

### Tests
- defines `baseParams ` to retrieve default value of `ConstructorParams` struct.
- add `test_RevertWhen_CampaignNameTooLong` to check for `name` length
- update `createMerkleStreamerLL.tree`

andreivladbrg commented on test/utils/Defaults.sol:
```diff
@@ -55,6 +55,8 @@ contract Defaults is Merkle {
     uint256[] public LEAVES = new uint256[](RECIPIENTS_COUNT);
     bytes32 public immutable MERKLE_ROOT;
+    string public constant NAME_STRING = "Airdrop Campaign";
+    bytes32 public constant NAME = 0x41697264726f702043616d706169676e00000000000000000000000000000000;
```
let's change to `bytes32(abi.encodePacked("Airdrop Campaign"))`

smol-ninja replied:
I agree.

andreivladbrg reviewed (changes requested):
Were you getting the `stack too deep` error in both `MerkleStreamer.constructor` and `MerkleStreamerFactory.create` function? 

Also, why do we pass a string in the constructor instead of `bytes32`, wouldn't it be more efficient?

> adds a new private storage _NAME of type bytes32 to store  name.

smart way of doing this, congrats

I've created a [separate branch](https://github.com/sablier-labs/v2-periphery/tree/feat/add-name) where I approached things differently. The reason I did not start from your branch is that it would have taken me more time to make the changes.

Could you take a look at my branch and tell me if you consider that approach better? If yes, I will polish it. I personally prefer it because the campaign contract is not modified as much.

PaulRBerg commented:
Thanks for the PR, Shub.

It's a shame that this additional parameter forces us to use a struct in the constructor. Structs have caused all sorts of problems for V2 integrators (e.g., [Tally](https://docs.tally.xyz/knowledge-base/proposals/creating-proposals/custom-actions/streaming-payments-with-sablier)).

But I can't see any way out - unless we remove another parameter, which we can't do.

My initial reaction was to suggest you discard this feature, but then I remembered LockupTranched/ LockupDynamic. It is very likely that the implementation of these other contracts will require a struct (because of the segments). So introducing structs in `MerkleStreamerLL` is fine (and this PR remains open).

andreivladbrg commented:
> other contracts will require a struct (because of the segments

haven't we decided that the segments would be passed in the claim function, but not in the create factory function?

https://github.com/sablier-labs/v2-periphery/discussions/166#discussioncomment-8123208

PaulRBerg commented:
Oooh, you're right, @andreivladbrg. But at any rate, I guess that other parameters will be needed in the constructor for these other implementations.

Anyway. I am happy to proceed with the implementation brought about by this PR.

andreivladbrg commented:
@PaulRBerg 

have you checked the [other branch](https://github.com/sablier-labs/v2-periphery/tree/feat/add-name)?

which version do you say it's better?

PaulRBerg commented:
I haven't checked the other branch, no. In fact, I haven't even checked this PR properly. I've just seen the struct and wanted to comment on that.

I will review it later.

smol-ninja commented on test/Base.t.sol:
```diff
@@ -245,6 +246,28 @@ abstract contract Base_Test is DeployOptimized, Events, Merkle, V2CoreAssertions
                                   MERKLE-STREAMER
     //////////////////////////////////////////////////////////////////////////*/
 
+    function createWithLockupLinear(
```
This can't be placed in `Defaults` because it requires access to `lockupLinear` and `asset` variables which are defined in `Base_Test`.

```solidity
    function createWithLockupLinear(
        address admin,
        bytes32 merkleRoot,
        uint40 expiration
    )
        internal
        view
        returns (MerkleStreamer.CreateWithLockupLinear memory)
    {
        return MerkleStreamer.CreateWithLockupLinear({
            initialAdmin: admin,
            lockupLinear: lockupLinear,
            asset: asset,
            name: defaults.NAME_STRING(),
            merkleRoot: merkleRoot,
            expiration: expiration,
            streamDurations: defaults.durations(),
            cancelable: defaults.CANCELABLE(),
            transferable: defaults.TRANSFERABLE()
        });
    }
```

andreivladbrg replied:
See how i have handled it in my branch

smol-ninja replied:
OK I will have a look at the other branch tomorrow first thing in the morning.

smol-ninja commented on src/abstracts/SablierV2MerkleStreamer.sol:
```diff
@@ -80,6 +93,11 @@ abstract contract SablierV2MerkleStreamer is
     }
 
+    /// @inheritdoc ISablierV2MerkleStreamer
+    function NAME() external view override returns (string memory) {
```
I agree. Even I didn't like it that way.

smol-ninja commented on src/types/DataTypes.sol:
```diff
@@ -58,3 +60,29 @@ library Batch {
+    /// @param streamDurations The durations for each stream due to the recipient.
+    /// @param cancelable Indicates if each stream will be cancelable.
+    /// @param transferable Indicates if each stream NFT will be transferable.
+    struct CreateWithLockupLinear {
```
In future, we will have `createMerkleStreamerLD` and `createMerkleStreamerLT` functions. The constructor parameters of those functions will have a set of different variables than the ones in `createMerkleStreamerLL`.

For example,
1.  `createMerkleStreamerLL` has `ISablierV2LockupLinear lockupLinear` and `LockupLinear.Durations streamDurations`
2. `createMerkleStreamerLD` may have `ISablierV2LockupDynamic lockupDynamic` and `Range range` or `SegmentWithDelta[] segments`
3. `createMerkleStreamerLT` may have `ISablierV2LockupTranched lockupTranched` and `Segment`

That's why I called `CreateWithLockupLinear` instead of giving it a generic name such as `ConstructorParams`.

smol-ninja commented on src/types/DataTypes.sol:
```diff
@@ -58,3 +60,29 @@ library Batch {
+library MerkleStreamer {
+    /// @notice Struct encapsulating the {SablierV2MerkleStreamerLL} parameters for the
+    /// {SablierV2MerkleStreamerFactory.createMerkleStreamerLL}
+    /// function.
```
Same reason as above. In future, we will have structs encapsulating other functions such as `createMerkleStreamerLD` and `createMerkleStreamerLT`.

smol-ninja commented on src/interfaces/ISablierV2MerkleStreamer.sol:
```diff
@@ -29,6 +29,9 @@ interface ISablierV2MerkleStreamer is IAdminable {
     function ASSET() external returns (IERC20);
 
+    /// @notice The name of the Merkle streamer contract.
+    function NAME() external returns (string memory);
```
I agree

smol-ninja commented on src/libraries/Errors.sol:
```diff
@@ -25,4 +25,7 @@ library Errors {
     error SablierV2MerkleStreamer_StreamClaimed(uint256 index);
+
+    /// @notice Thrown when name of the Merkle Streamer contract is longer than 32 bytes.
+    error SablierV2MerkleStreamer_NameTooLong();
```
Agree.

smol-ninja commented on src/abstracts/SablierV2MerkleStreamer.sol:
```diff
@@ -53,13 +60,19 @@ abstract contract SablierV2MerkleStreamer is
         bool cancelable,
         bool transferable
     ) {
+        if (bytes(name).length > 32) {
```
I had the following reason to place this check in `SablierV2MerkleStreamer`:

We still need this check in `SablierV2MerkleStreamerLL` since that contract can independently be deployed. So ideally, the check on name length should be a part of this contract. However, since we import `Errors` in `SablierV2MerkleStreamer`, I decided to keep the check in `SablierV2MerkleStreamer` to avoids re-importing `Errors` in `SablierV2MerkleStreamerLL`.

I am happy to keep the check in `SablierV2MerkleStreamerFactory` as well as `SablierV2MerkleStreamerLL` but then it will just repeat itself.

andreivladbrg commented:
After rethinking about this, I consider now that it would be better to have a struct:
```solidity
```

andreivladbrg commented:
thanks for your responses @smol-ninja 

>I chose string because I was worried that integrators might make mistakes while converting the string into bytes as it follows a big-endian format in solidity.

this , and there is one more thing if we keep the check in the constructor, we won't need to make the same check in every `createMerklerStreamer` function. so i am happy to leave it there

After rethinking about this, I consider now that it would be better to have a struct that cointains only the parameters passed in the abstract contract:

```solidity
library MerkleStreamer {
    struct ConstructorParams {
        address initialAdmin;
        string name;
        IERC20 asset;
        bytes32 merkleRoot;
        uint40 expiration;
        bool cancelable;
        bool transferable;
    }
}
```

So that it can be used in all future contracts that inherit `MerkleStreamer`. Ofc, this needs to tested to see if it really worki in practic (if we no longer have `stack too deep` error):

```solidity
function createMerkleStreamerLL(
    MerkleStreamer.ConstructorParams memory params,
    ISablierV2LockupLinear lockupLinear,
    LockupLinear.Durations streamDurations,
    string ipfsCID,
    uint256 aggregateAmount,
    uint256 recipientsCount
)
    external
    returns (ISablierV2MerkleStreamerLL merkleStreamerLL); 
```

smol-ninja commented:
Great idea @andreivladbrg. I prefer this approach too. This way, we won't have to deal with separate structs for future streamer contracts. I think it should work because it would nonetheless reduce the no. of stack variables for the factory create function.

I would also like to know your comment on my 2nd point: "not having `ISablierV2LockupLinear` in `Defaults.sol`".

Whatever decision @PaulRBerg takes on discarding this feature or not, we should still do this because there might be new variables in the factory create versions of LD and LT that would eventually place us into `stack too deep` error. 

If you agree, I am happy to make these changes and update the PR.

andreivladbrg commented:
> I would also like to know your comment on my 2nd point: "not having ISablierV2LockupLinear in Defaults.sol

well, if we switch to the proposal i've made, we will no longer need the `ISablierV2LockupLinear` in `Defaults.sol`

smol-ninja commented:
@andreivladbrg I've updated the PR based on our newest proposal. I have also taken some suggestions from your branch.

andreivladbrg commented on test/utils/Defaults.sol:
```diff
@@ -114,6 +116,30 @@ contract Defaults is Merkle {
+        return createConstructorParams(users.admin, MERKLE_ROOT, EXPIRATION);
+    }
+
+    function createConstructorParams(
```
Let's say only `cosntructorParams`

andreivladbrg reviewed (approved):
looking good now

>might be new variables in the factory create versions of LD and LT

well, what are these new variables? maybe they will even be less, we will remove the `LockupLinear.Durations` struct, but we will add `MAX_SEGMENT_COUNT`

smol-ninja commented:
> maybe they will even be less, we will remove the LockupLinear.Durations struct, but we will add MAX_SEGMENT_COUNT

Yes, you are right.

PaulRBerg commented on src/SablierV2MerkleStreamerFactory.sol:
```diff
@@ -36,33 +31,53 @@ contract SablierV2MerkleStreamerFactory is ISablierV2MerkleStreamerFactory {
+    //////////////////////////////////////////////////////////////////////////*/
 
+    /// @dev Helper function to emit the {CreateMerkleStreamerLL} event.
+    function _emitLLEvent(
```
Why are we not simply emitting the `params` struct in the event instead of doing this?

It should be pretty easy to parse that struct in the subgraph (since the subgraph has to be able to parse structs anyway).

Cc @razgraf can lend us a hand here?

razgraf replied:
I guess we could, are there any variables we're not emitting (for gas purposes) in the iterative way that we would by replacing it with the params structs?

andreivladbrg replied:
imo, here is it not about gas, but about DX

would it affect the dev work?

razgraf replied:
We're adding a new parameter anyway so the event has to be duplicated regardless. Of course, changing from a list into a single struct parameter will make union types e.g. `event: Event1 | Event2` harder to process but it's a minor point in the grand scheme.

andreivladbrg replied:
ok, got it

then, we will emit the struct

PaulRBerg commented on src/abstracts/SablierV2MerkleStreamer.sol:
```diff
@@ -50,20 +58,22 @@ abstract contract SablierV2MerkleStreamer is
+    constructor(MerkleStreamer.ConstructorParams memory params) {
+        // Checks: the campaign name is not greater than 32 bytes
+        if (bytes(params.name).length > 32) {
+            revert Errors.SablierV2MerkleStreamerFactory_CampaignNameTooLong({
```
This error is not thrown in the `Factory` contract, so it should be renamed.

PaulRBerg commented on test/integration/merkle-streamer/factory/create-merkle-streamer-ll/createMerkleStreamerLL.t.sol:
```diff
@@ -12,26 +14,47 @@ contract CreateMerkleStreamerLL_Integration_Test is MerkleStreamer_Integration_T
+            abi.encodeWithSelector(
+                Errors.SablierV2MerkleStreamerFactory_CampaignNameTooLong.selector, bytes(params.name).length, 32
+            )
+        );
```
We should also write a fuzz test that fuzzes the `name`.

Or maybe we can incorporate this fuzzing in the existing `testFuzz_CreateMerkleStreamerLL` test?

smol-ninja replied:
Interesting, but why do you want to fuzz `name`? We have an [integration test for name length](https://github.com/sablier-labs/v2-periphery/blob/af89263a36675a30efb1bcb2e2ec14a27c25003c/test/integration/merkle-streamer/factory/create-merkle-streamer-ll/createMerkleStreamerLL.t.sol#L17C1-L40C6) and since it is constant, as long as it fits into 32 bytes, there can never be a problem.

PaulRBerg reviewed (changes requested):
I like your approach to modularize the constructor params in (i) shared params and (ii) bespoke params for each shape (LL, LT, etc.)

However:

- I would prefer the name of the struct parameter to better reflect the fact that the parameters are shared. Maybe we could rename it to `baseParams`, `sharedParams`, `commonParams`, or something along these lines. Suggestions welcome.
- Why are `ipfsCID`, `aggregateAmount`, and `recipientsCount` not part of the `ConstructorParams` struct? They seem to be common to all campaigns, not just the LL campaigns.

andreivladbrg commented:
> Why are ipfsCID, aggregateAmount, and recipientsCount not part of the ConstructorParams struct? They seem to be common to all campaigns

because these parameters **are not** passed in the constructor and the struct passed is called `ConstructorParams` and it's within `MerkleStreamer` namespace, **not** _Factory_:

https://github.com/sablier-labs/v2-periphery/blob/17d8d942d28dd6329129550894eb418df9029549/src/abstracts/SablierV2MerkleStreamer.sol#L61

do you want to add a new library `MerkleStreamerFactory` as i did in my [separate branch](https://github.com/sablier-labs/v2-periphery/blob/fc6279917d8152fc20f4c00fe1bd79c8b6fb6ea0/src/types/DataTypes.sol#L64-L90), which you can also review?

initially considered the same thing as being better, but as i said in my comment [here](https://github.com/sablier-labs/v2-periphery/pull/262#issuecomment-1898478184), i no longer find it to be the better option

andreivladbrg commented on src/abstracts/SablierV2MerkleStreamer.sol:
```diff
@@ -38,6 +39,13 @@ abstract contract SablierV2MerkleStreamer is
+    //////////////////////////////////////////////////////////////////////////*/
+
+    /// @dev The name of the campaign stored as bytes32.
+    bytes32 private immutable _NAME;
```
i guess `private` is more gas efficient, and we don't need this variable anywhere else

PaulRBerg commented:
> because these parameters **are not** passed in the constructor

Got it. Makes sense.

My feedback about the name of the struct parameter still stands, though.

Can we rename it to `baseParams` or `commonParams`?

> Do you want to add a new library `MerkleStreamerFactory`

No.

smol-ninja commented on src/abstracts/SablierV2MerkleStreamer.sol:
```diff
@@ -50,20 +58,22 @@ abstract contract SablierV2MerkleStreamer is
+        }
+
+        admin = params.initialAdmin;
+        ASSET = params.asset;
+        _NAME = bytes32(abi.encodePacked(params.name));
+        MERKLE_ROOT = params.merkleRoot;
+        EXPIRATION = params.expiration;
+        CANCELABLE = params.cancelable;
+        TRANSFERABLE = params.transferable;
```
Sure. I will also rename `_NAME` to `NAME`

smol-ninja commented:
I agree with @andreivladbrg on renaming `ConstructorParams`. I don't see any of them more relevant than `ConstructorParams` other than the fact that they are shorter names. I also don't like `ConstructorParams` but unfortunately I couldn't think of a better name that can express all the parameters of the `MerkleStreamer` constructor.

smol-ninja commented:
@PaulRBerg do you have any suggestions on this? Andrei has already approved the changes.

andreivladbrg commented:
shouldn't `baseParams` name be only in the `createMerkleStreamerLL` function?

why does it need to be named "base" in the abstract contract as well?

smol-ninja commented:
Yes! You're right @andreivladbrg. That was a mistake.

PaulRBerg reviewed (approved):
LGTM now

PaulRBerg commented:
Though the build fails in CI

smol-ninja commented:
That's because we updated [package.json](https://github.com/sablier-labs/v2-periphery/blob/staging/package.json#L15) in https://github.com/sablier-labs/v2-periphery/pull/264 without updating bun lockfile. Fixed it now.

```json
"@sablier/v2-core": "github:sablier-labs/v2-core#staging"
```
